compiler-core/src/language_server/engine.rs

```diff
@@ -543,3 +582,41 @@ fn hover_for_expression(expression: &TypedExpr, line_numbers: LineNumbers) -> Ho
         range: Some(src_span_to_lsp_range(expression.location(), &line_numbers)),
     }
 }
+
+fn is_action_attached_to_unused_diagnostic(params: &lsp::CodeActionParams) -> bool {
+    params
+        .context
+        .diagnostics
+        .iter()
+        .any(|diag| diag.message.starts_with("Unused"))
```

Comment:
Check this from the `unused_imports` vector please. Diagnostic messages are not a reliable interface, and it's possible this buffer has already been flushed at this point.

Reply:
Right, so for this part I'm not entirely sure how LSP works. If we remove this check, and use the `unused_imports` vector, then the server will always suggest to remove unused imports, even when the click on unrelated warnings. Another solution would be to check if the code action request params range falls into an import from the ast.

One thing I couldn't find yet, is how to suggest code action without waiting for the user request, I guess that would be the right thing to do instead of a waiting for a code action request.

Reply:
It must not be based off of warnings at all. They are not an interface for any kind of reflection.

From reading the code it adds the actions for each item in the `unused_imports` vector, so if there's none then none will be added.

Reply:
Ok, but if there is an unused import, then the action will be proposed, even when the user click on an unrelated warning. May I lookup the action request location to see if it match an import location?

Reply:
Sorry, I don't understand. Why would the `unused_imports` vector be filled with locations of unrelated warnings?

Reply:
Oh no, the `unused_imports` only contains the locations of unused imports. What I am trying to say is that presently, the LanguageServerEngine.action method is called when the user requests a code action and this [request](https://docs.rs/lsp-types/latest/lsp_types/struct.CodeActionParams.html) can come from anywhere in the document.

Therefor, only checking the `unused_imports` vector is not enough, we need to make sure the user actually request a code action that is coming from an import location, and not, for example, from a non exhaustive pattern match error.

Currently the server always offers to remove unused import (if there are any), but we could improve that by checking that the request range is overlapping with import node from the ast, that way, the action would be only available when the user clicks on an unused import warning.

Hope that is more clear.

Reply:
Yes, that's right. The user presses the hotkey for suggestions for code actions, the editor asks the server, the server returns a code action if the position is in the `unused_imports`, so the user only gets a code action suggested if they are positioned in an unused value.

Reply:
Oh right, we can use the location of the unused_imports directly instead of the looking up the ast. I'll give this a try shortly.

Sorry for the confusion, I got a bit lost here :)

Reply:
Alright, this is now implemented.